common/lru/blob_lru.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package lru
+
+import (
+	"math"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+)
+
+// SizeConstrainedLRU is a wrapper around simplelru.LRU. The simplelru.LRU is capable
+// of item-count constraints, but is not capable of enforcing a byte-size constraint,
+// hence this wrapper.
+// OBS: This cache assumes that items are content-addressed: keys are unique per content.
+// In other words: two Add(..) with the same key K, will always have the same value V.
+type SizeConstrainedLRU struct {
+	size    uint64
+	maxSize uint64
+	lru     *simplelru.LRU
+	lock    sync.RWMutex
+}
+
+// NewSizeConstraiedLRU creates a new SizeConstrainedLRU.
+func NewSizeConstraiedLRU(max uint64) *SizeConstrainedLRU {
+	lru, err := simplelru.NewLRU(math.MaxInt, nil)
+	if err != nil {
+		panic(err)
+	}
+	return &SizeConstrainedLRU{
+		size:    0,
+		maxSize: max,
+		lru:     lru,
+	}
+}
+
+// Set adds a value to the cache.  Returns true if an eviction occurred.
+// OBS: This cache assumes that items are content-addressed: keys are unique per content.
+// In other words: two Set(..) with the same key K, will always have the same value V.
+func (c *SizeConstrainedLRU) Set(key []byte, value []byte) (evicted bool) {
```

Comment:
The only difference between Set and Add is they have different parameter types? Can we remove one of them maybe and only have a single API?

Reply:
Yes.. I added `Set([]byte, []byte)` because that's what the previous one (the one we're replacing) used, so it minimized the diff. I can remove that one if you prefer

Reply:
I removed one of them, and changed the signature a bit, so it uses `common.Hash` for key (add/get)